src/WebJobs.Script.WebHost/Controllers/HostController.cs

```diff
@@ -154,7 +163,57 @@ public async Task<IActionResult> SyncTriggers()
         public IActionResult Restart([FromServices] IScriptHostManager hostManager)
         {
             Task ignore = hostManager.RestartHostAsync();
-            return Ok(_webHostSettings.Value);
+            return Ok(_applicationHostOptions.Value);
+        }
+
+        /// <summary>
+        /// Currently this endpoint only supports taking the host offline and bringing it back online.
+        /// </summary>
+        /// <param name="state">The desired host state. See <see cref="ScriptHostState"/>.</param>
+        [HttpPut]
+        [Route("admin/host/state")]
+        [Authorize(Policy = PolicyNames.AdminAuthLevel)]
+        public async Task<IActionResult> SetState()
+        {
+            if (string.Compare(Request.ContentType, "text/plain", StringComparison.OrdinalIgnoreCase) != 0)
+            {
+                return BadRequest();
+            }
+
+            string state;
+            using (StreamReader sr = new StreamReader(Request.Body, Encoding.UTF8))
+            {
+                state = await sr.ReadToEndAsync();
+            }
+
+            if (!Enum.TryParse<ScriptHostState>(state, ignoreCase: true, out ScriptHostState desiredState))
+            {
+                return BadRequest();
+            }
+
+            var currentState = _scriptHostManager.State;
+            if (desiredState == ScriptHostState.Running && currentState == ScriptHostState.Offline)
+            {
+                // we're currently offline and the request is to bring the host back online
+                FileMonitoringService.SetAppOfflineState(_applicationHostOptions.Value.ScriptPath, false);
+            }
+            else if (desiredState == ScriptHostState.Offline && currentState != ScriptHostState.Offline)
+            {
+                if (_environment.FileSystemIsReadOnly())
+                {
+                    // taking the host offline is based on app_offline.htm file creation
+                    return BadRequest();
+                }
+
+                // we're currently online and the request is to take the host offline
+                FileMonitoringService.SetAppOfflineState(_applicationHostOptions.Value.ScriptPath, true);
+            }
+            else
+            {
+                return BadRequest();
```

Comment:
This is a HTTP PUT so you should allow idempotent operations (e.g. where the desired state matches the current state).

Also for the else case, maybe return a string such as "transistion from current state {x} to requested state {y} is not supported" as part of the bad request?

Reply:
Fixed. For operations where desiredState == currentState, I just return OK. 400 is now only returned for invalid state transitions.